### ISSUE TITLE
.Site.BaseURL is not available in one context, but not another

### ISSUE THREAD

**Issue by kevingoldsmith** [spf13/hugo#751]
Getting the error: "BaseURL is not a field of struct type *hugolib.SiteInfo" when I use it in one template, as such:
```
{{ partial "header.html" . }}

<h1><a href="{ .Site.BaseURL }}/artists/{{ .Params.artistdirectory }}">{{ .Params.artistname }}</a> - {{ .Params.releasename }}</h1>
Released: {{ .Params.releaseyear }}<br />
{{ .Content }}
```

But in this peer template, it works fine:

```
{{ partial "header.html" . }}

		<h1>{{ .Params.artistname }}</h1>
		{{ .Content }}

		{{ if isset .Params "albums" }}
			<h1>Releases</h1>
			{{ range index .Params "albums" }}
				<a href="{{ .Site.BaseURL }}/releases/{{ .catalogid }}">{{ .name }}</a>
			{{ end }}
		{{ end }}

{{ partial "footer.html" . }}
```

The only thing I can think of is that for some reason, it isn't available in the higher context, but it the deeper one.

**Comment by bjornerik** [comment]
Hey,

This is Go templates, and I doubt there is a bug in there about this, but it's hard to test your case without a complete sample (test project).

I have experienced this myself, and I'm pretty sure this is related to some "template oddities" I've experienced myself. I will take an commented example of "wrong", "looks right" and "is right":

Wrong:

```
{{ with .Params.myparam }}
{{ .Site.BaseURL }} // the "." is now at myparam so this will fail
{{ partial "mypartial.html" . }} // again, wrong "." sent in param to partial
{{ end }}
```

Looks right:
```
{{ $ctx := . }}
{{ with .Params.myparam }}
{{ $ctx.Site.BaseURL }} // this should work fine
{{ partial "mypartial.html" . }} // easy to miss
{{ end }}
```

Is right:

```
{{ $ctx := . }}
{{ with .Params.myparam }}
{{ $ctx.Site.BaseURL }}
{{ partial "mypartial.html" $ctx }}
{{ end }}
```

Another pitfall that took me some time to figure out (that I have a pull request in to compensate for, see #746), is the visibility of variables:

```
{{ $var := 1 }}
{{ $ctx := . }}
{{ with .Params.myparam }}
{{ $var := 2 }}
{{ partial "mypartial.html" $ctx }} // $var will not be available to mypartial.html 
{{ end }}
{{ $var }} // => 1
```
@anthonyfok some of these common pitfalls should have a place in the documentation. I'm not sure that all of the above is 100% runnable code, but it shows the point.

**Comment by kevingoldsmith** [comment]
Thanks for your continuing help. At the moment I am just working around this so that I can finish the project I am working on. Once that is done, I will try to clean this up and either come back with a better case I can put on github or I will figure out what my mental block is with Go scoping rules :)

Warning: may contain auto-correct frijoles.

**Comment by bjornerik** [comment]
OK, please reopen this or open another issue when that time comes.

**Comment by abourget** [comment]
Is it possible that this issue is related to a `{{partial "location" .SomeVariable}}`.. and then from within the partial, `$.Site` or `.Site` is not available anymore ? Passing `{{partial "location" . }}` keeps the same variables within the partial, but anything other than `.` as a second parameter will "out-scope" variables, `$.` included.

I've hit an issue that seems like this one, can you confirm my understanding ? If it is such, I'll add some docs to https://gohugo.io/templates/partials/
